sql-plugin/src/main/scala/com/nvidia/spark/rapids/spill/SpillFramework.scala

```diff
@@ -1597,59 +1591,72 @@ object SpillFramework extends Logging {
  *
- * `release` actually closes the underlying DeviceMemoryBuffer, and should be called
+ * `release` actually closes the underlying buffer, and should be called
  * once at the end of the lifetime of the executor.
  *
- * @param dmb - actual cudf DeviceMemoryBuffer that this class is protecting.
+ * @param buf - actual cudf DeviceMemoryBuffer that this class is protecting.
+ * @param pool - the pool to which this buffer belongs
  */
-private[spill] case class DeviceBounceBuffer(var dmb: DeviceMemoryBuffer) extends AutoCloseable {
-  private var acquired: Boolean = false
-  def acquire(): DeviceBounceBuffer = synchronized {
-    while (acquired) {
-      wait()
-    }
-    acquired = true
-    this
-  }
-
-  private def unaquire(): Unit = synchronized {
-    acquired = false
-    notifyAll()
-  }
-
+private[spill] class BounceBuffer[T <: AutoCloseable](
+                                                       var buf: T,
+                                                       private val pool: BounceBufferPool[T])
+  extends AutoCloseable {
   override def close(): Unit = {
-    unaquire()
+    pool.returnBuffer(this)
   }
 
-  def release(): Unit = synchronized {
-    if (acquired) {
-      throw new IllegalStateException(
-        "closing device buffer pool, but some bounce buffers are in use.")
-    }
-    if (dmb != null) {
-      dmb.close()
-      dmb = null
-    }
+  def release(): Unit = {
+    buf.close()
+    buf = null.asInstanceOf[T]
   }
 }
 
+
 /**
- * A bounce buffer pool with buffers of size `bufferSize`
+ * A bounce buffer pool with buffers of size `bufSize`
  *
- * This pool returns instances of `DeviceBounceBuffer`, that should
+ * This pool returns instances of `BounceBuffer[T]`, that should
  * be closed in order to be reused.
  *
  * Callers should synchronize before calling close on their `DeviceMemoryBuffer`s.
  */
-trait DeviceBounceBufferPool extends AutoCloseable {
-  def bufferSize: Long
-  def nextBuffer(): DeviceBounceBuffer
+class BounceBufferPool[T <: AutoCloseable](private val bufSize: Long,
+                                           private val bbCount: Long,
+                                           private val allocator: Long => T)
+  extends AutoCloseable with Logging {
+
+  private val pool = new LinkedBlockingQueue[BounceBuffer[T]]
+  for (_ <- 1L to bbCount) {
+    pool.offer(new BounceBuffer[T](allocator(bufSize), this))
+  }
+
+  def bufferSize: Long = bufSize
+  def nextBuffer(): BounceBuffer[T] = pool.take()
+  def returnBuffer(buffer: BounceBuffer[T]): Unit = {
+    if (closed) {
+      buffer.release()
+    } else {
+      pool.offer(buffer)
+    }
+  }
+
+  private var closed = false
+  override def close(): Unit = {
+    if (!closed) {
+      closed = true
+      if (pool.size() < bbCount) {
+        logError("tried to close BounceBufferPool when buffers are still " +
+          "being used")
+      }
+
+      pool.forEach(_.release())
```

Comment:
we should remove all bbs from the pool in the close() method under a synchronized block to protect against `nextBuffer` getting a released bb.